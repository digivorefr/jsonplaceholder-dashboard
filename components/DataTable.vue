<template>
  <div class="data-table">
    <div class="data-table__scroller">
      <div class="data-table__wrapper">
        <div
          v-for="columnIndex in columns"
          :key="columnIndex"
          class="data-table__column"
        >
          <div
            v-if="headers !== null"
            class="data-table__cell data-table__cell--heading"
          >
            {{ headers[columnIndex -1] }}
          </div>
          <DataTableCell
            v-for="(row, index) in values"
            :key="index"
            :value="row[columnIndex-1]"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue, { VNode } from 'vue'
import DataTableCell, { Value } from '~/components/DataTableCell.vue'
// export type Value = string | VNode | null;
export type Action = VNode | null;

export type Options = {
  values: Value[][],
  headers?: string[],
}

export default Vue.extend({
  name: 'DataTable',
  components: { DataTableCell },
  props: {
    options: {
      type: Object as () => Options,
      required: true
    }
  },
  computed: {
    headers(): string[] | null {
      if (this.options.headers !== undefined &&
        this.options.headers !== null &&
        this.options.headers.length !== this.options.values[0].length) {
        this.$nuxt.error({ statusCode: 500, message: 'DataTable headers and values must have the same length.' })
      }
      return this.options.headers || null
    },
    values(): Value[][] {
      return this.options.values
    },
    columns(): number { return (this.values[0] || []).length },
    rows(): number { return this.values.length }
  }
})
</script>
