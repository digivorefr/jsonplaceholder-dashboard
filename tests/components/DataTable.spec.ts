import { shallowMount, Wrapper } from '@vue/test-utils'
import DataTable from '@/components/DataTable.vue'

const entriesMock = {
  headers: ['Name', 'Role'],
  values: [
    ['Marty', 'Hero'],
    ['Emmett', 'Scientist'],
    ['Biff', 'Vilain']
  ]
}

describe('DataTable', () => {
  test('renders correctly', () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(DataTable, {
      propsData: {
        options: {
          headers: null,
          values: []
        }
      }
    })
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.vm.columns).toBe(0)
    wrapper.destroy()
  })

  test('table well formed', () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(DataTable, {
      propsData: {
        options: entriesMock
      }
    })
    expect(wrapper.vm.columns).toBe(2)
    expect(wrapper.vm.rows).toBe(3)
    expect(wrapper.find('.data-table__cell--heading'))
    expect(wrapper.element).toMatchSnapshot()
    wrapper.destroy()
  })
  test('table well formed without headers', () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(DataTable, {
      propsData: {
        options: { ...entriesMock, headers: null }
      }
    })
    expect(wrapper.vm.headers).toBe(null)
    wrapper.destroy()
  })

  test('configuration mismatch between headers and values', () => {
    const nuxtMock = jest.fn()

    const wrapper: Wrapper<Vue & { [key: string]: any }> = shallowMount(DataTable, {
      propsData: {
        options: {
          ...entriesMock,
          headers: ['Back to the future']
        }
      },
      mocks: {
        $nuxt: { error: nuxtMock }
      }
    })
    expect(nuxtMock).toHaveBeenCalledWith({ statusCode: 500, message: 'DataTable headers and values must have the same length.' })
    wrapper.destroy()
  })
})
