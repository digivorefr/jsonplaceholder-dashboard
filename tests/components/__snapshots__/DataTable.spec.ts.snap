// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable table well formed 1`] = `
<div
  class="data-table"
>
  <div
    class="data-table__scroller"
  >
    <div
      class="data-table__wrapper"
    >
      <div
        class="data-table__column"
      >
        <div
          class="data-table__cell data-table__cell--heading"
        >
          
          Name
        
        </div>
         
        <datatablecell-stub
          value="Marty"
        />
        <datatablecell-stub
          value="Emmett"
        />
        <datatablecell-stub
          value="Biff"
        />
      </div>
      <div
        class="data-table__column"
      >
        <div
          class="data-table__cell data-table__cell--heading"
        >
          
          Role
        
        </div>
         
        <datatablecell-stub
          value="Hero"
        />
        <datatablecell-stub
          value="Scientist"
        />
        <datatablecell-stub
          value="Vilain"
        />
      </div>
    </div>
  </div>
</div>
`;
